File: algorithms/python/selection_sort.py
from typing import MutableSequence


def swap(l: MutableSequence[int], i1: int, i2: int) -> None:
  l[i1], l[i2] = l[i2], l[i1]


def selection_sort_v1(l: MutableSequence[int]) -> None:
  for i in range(len(l) - 1):
    smallest = i
    for j in range(i + 1, len(l)):
      if l[j] < l[smallest]:
        smallest = j
    if smallest != i:
      swap(l, i, smallest)


a = [3, 1, 2, 5, 4]

print(a)
selection_sort_v1(a)
print(a)

File: algorithms/C/selection_sort.c
#include <stdio.h>


typedef unsigned int ui32;


static void swap(int *array, ui32 i1, ui32 i2) {
  int temp = *(array + i1);
  *(array + i1) = *(array + i2);
  *(array + i2) = temp;
}


static void selection_sort(int *array, ui32 len) {
  ui32 smallest;
  for (ui32 i = 0; i < len - 1; i++) {
    smallest = i;
    for (ui32 j = i + 1; j < len; j++) {
      if (array[j] < array[smallest]) {
        smallest = j;
      }
    }
    if (smallest != i) {
      swap(array, i, smallest);
    }
  }
}


static void print_array(int *array, ui32 len) {
  for (ui32 i = 0; i < len; i++) {
    printf("%d ", array[i]);
  }
  printf("\n");
}


int main(void) {
  int array[5] = { 3, 1, 2, 5, 4 };
  print_array(array, 5);
  selection_sort(array, 5);
  print_array(array, 5);
  return 0;
}
